File: python/src/domain/market/convert.py
"""Market wire-to-domain conversion."""

from typing import Optional
from . import (
    Market, Status, Outcome, ConditionalToken, DepositAsset,
    OrderBookPairSummary, TokenMetadata,
)
from .wire import MarketWire


def _parse_status(s: Optional[str]) -> Status:
    if s is None:
        return Status.PENDING
    return Status.from_str(s)


def validation_errors_from_wire(wire: MarketWire) -> list[str]:
    errors: list[str] = []

    if not wire.slug:
        errors.append("Missing slug")
    if not wire.market_name:
        errors.append("Missing name")
    if not wire.description:
        errors.append("Missing description")
    if not wire.definition:
        errors.append("Missing definition")
    if not wire.icon_url:
        errors.append("Missing thumbnail image")
    if not wire.banner_image_url:
        errors.append("Missing banner image")
    if wire.market_status and wire.market_status not in {"Pending", "Active", "Resolved", "Cancelled"}:
        errors.append("Invalid status")

    if not errors:
        return []

    identifier = wire.market_pubkey or str(wire.market_id)
    return [f"Market validation errors ({identifier}): {', '.join(errors)}"]


def market_from_wire(wire: MarketWire) -> Market:
    """Convert a MarketWire to a Market domain type."""
    outcomes = [
        Outcome(
            index=o.get("index", i),
            name=o.get("name", ""),
            icon_url=o.get("icon_url", ""),
        )
        for i, o in enumerate(wire.outcomes)
    ]

    # Build conditional tokens and token_metadata from deposit_assets wire
    conditional_tokens: list[ConditionalToken] = []
    token_metadata: dict[str, TokenMetadata] = {}
    deposit_assets: list[DepositAsset] = []
    orderbook_ids: list[str] = []

    for da in wire.deposit_assets:
        deposit_asset = DepositAsset(
            id=da.get("id", 0),
            market_pda=da.get("market_pubkey", ""),
            deposit_asset=da.get("deposit_asset", ""),
            num_outcomes=da.get("num_outcomes", 0),
            name=da.get("display_name", ""),
            symbol=da.get("token_symbol", da.get("symbol", "")),
            description=da.get("description"),
            decimals=da.get("decimals") or 6,
            icon_url=da.get("icon_url", ""),
        )
        deposit_assets.append(deposit_asset)

        for ct in da.get("conditional_mints", []):
            conditional_tokens.append(ConditionalToken(
                mint=ct.get("token_address", ""),
                outcome_index=ct.get("outcome_index", 0),
                outcome=ct.get("outcome", ""),
                deposit_asset=da.get("deposit_asset", ""),
                deposit_symbol=da.get("token_symbol", da.get("symbol", "")),
                name=ct.get("display_name", ct.get("name", "")),
                symbol=ct.get("short_name", ct.get("symbol", "")),
                description=ct.get("description"),
                decimals=ct.get("decimals") or 6,
                icon_url=ct.get("icon_url", ""),
            ))
            mint = ct.get("token_address", "")
            if mint:
                token_metadata[mint] = TokenMetadata(
                    pubkey=mint,
                    symbol=ct.get("short_name", ct.get("symbol", "")),
                    decimals=ct.get("decimals") or 6,
                    icon_url=ct.get("icon_url", ""),
                    name=ct.get("display_name", ct.get("name", "")),
                )

    orderbook_pairs = []
    for ob in wire.orderbooks:
        ob_id = ob.get("orderbook_id", "")
        orderbook_pairs.append(OrderBookPairSummary(
            id=ob.get("id", 0),
            market_pubkey=ob.get("market_pubkey", wire.market_pubkey),
            orderbook_id=ob_id,
            base_token=ob.get("base_token", ""),
            quote_token=ob.get("quote_token", ""),
            outcome_index=ob.get("outcome_index", 0),
            tick_size=ob.get("tick_size", 0),
            total_bids=ob.get("total_bids", 0),
            total_asks=ob.get("total_asks", 0),
            last_trade_price=ob.get("last_trade_price"),
            last_trade_time=ob.get("last_trade_time"),
            active=ob.get("active", True),
        ))
        if ob_id:
            orderbook_ids.append(ob_id)

    return Market(
        id=wire.market_id,
        pubkey=wire.market_pubkey,
        name=wire.market_name,
        banner_image_url=wire.banner_image_url or "",
        icon_url=wire.icon_url or "",
        featured_rank=wire.featured_rank,
        volume=wire.volume or "0",
        slug=wire.slug or "",
        status=_parse_status(wire.market_status),
        created_at=wire.created_at,
        activated_at=wire.activated_at,
        settled_at=wire.settled_at,
        winning_outcome=wire.winning_outcome,
        description=wire.description or "",
        definition=wire.definition or "",
        category=wire.category,
        tags=wire.tags,
        deposit_assets=deposit_assets,
        conditional_tokens=conditional_tokens,
        outcomes=outcomes,
        orderbook_pairs=orderbook_pairs,
        orderbook_ids=orderbook_ids,
        token_metadata=token_metadata,
    )

File: typescript/src/domain/market/convert.ts
import { asPubkeyStr } from "../../shared";
import { MAX_OUTCOMES, MIN_OUTCOMES } from "../../program/constants";
import type { OrderBookPair } from "../orderbook";
import { orderBookPairFromWire } from "../orderbook/convert";
import { resolveIconUrls } from "./icon";
import { outcomeFromWire } from "./outcome";
import { statusFromWire, type Market, MarketValidationError, Status } from "./index";
import {
  globalDepositAssetFromWire,
  sortByDisplayPriority,
  validatedTokensFromWire,
  type DepositAsset,
  type DepositAssetPair,
} from "./tokens";
import type { MarketResponse } from "./wire";

export { globalDepositAssetFromWire, resolveIconUrls };

export function marketFromWire(source: MarketResponse): Market {
  const errors: string[] = [];
  const numOutcomes = source.num_outcomes ?? source.deposit_assets[0]?.num_outcomes ?? 0;
  if (!Number.isInteger(numOutcomes) || numOutcomes < MIN_OUTCOMES || numOutcomes > MAX_OUTCOMES) {
    errors.push(`Invalid outcome count: ${numOutcomes}`);
  }
  const inconsistentOutcomeCounts = source.deposit_assets
    .map((asset) => asset.num_outcomes)
    .filter((count) => count !== numOutcomes);
  if (inconsistentOutcomeCounts.length > 0) {
    errors.push(`Deposit asset outcome counts do not match market: ${inconsistentOutcomeCounts.join(", ")}`);
  }

  const outcomes = source.outcomes.flatMap((outcome) => {
    try {
      return [outcomeFromWire(outcome)];
    } catch (error) {
      errors.push(error instanceof Error ? error.message : String(error));
      return [];
    }
  });

  const depositAssets = [] as Market["depositAssets"];
  const conditionalTokens = [] as Market["conditionalTokens"];
  const tokenMetadata: Market["tokenMetadata"] = {};

  for (const depositAsset of source.deposit_assets) {
    try {
      const validated = validatedTokensFromWire(depositAsset);
      depositAssets.push(validated.token);
      conditionalTokens.push(...validated.conditionals);
      Object.assign(tokenMetadata, validated.metadata);
    } catch (error) {
      errors.push(error instanceof Error ? error.message : String(error));
    }
  }

  const orderbookPairs = source.orderbooks.flatMap((orderbook) => {
    try {
      return [orderBookPairFromWire(orderbook, conditionalTokens)];
    } catch (error) {
      errors.push(error instanceof Error ? error.message : String(error));
      return [];
    }
  });

  const status = statusFromWire(source.market_status);
  const definition = typeof source.definition === "string" ? source.definition : undefined;
  if (!source.slug) errors.push("Missing slug");
  if (!source.market_name) errors.push("Missing market name");
  if (!status) errors.push(`Invalid status: ${source.market_status}`);
  if (!definition) errors.push("Missing definition");

  const iconUrls = resolveIconUrls(source.icon_url_low, source.icon_url_medium, source.icon_url_high);
  if (!iconUrls) errors.push("Missing icon URL");

  // Banners are optional; cross-fallback still applies when any variant is set.
  const bannerUrls = resolveIconUrls(source.banner_image_url_low, source.banner_image_url_medium, source.banner_image_url_high);

  const depositAssetPairs = sortByDisplayPriority(
    deriveDepositAssetPairs(depositAssets, orderbookPairs),
  );

  if (depositAssetPairs.length === 0) {
    errors.push("Missing deposit asset pairs");
  }

  if (errors.length > 0) {
    throw new MarketValidationError(source.market_pubkey, errors);
  }

  return {
    id: source.market_id,
    pubkey: asPubkeyStr(source.market_pubkey),
    name: source.market_name ?? "",
    bannerImageUrlLow: bannerUrls?.low,
    bannerImageUrlMedium: bannerUrls?.medium,
    bannerImageUrlHigh: bannerUrls?.high,
    iconUrlLow: iconUrls?.low ?? "",
    iconUrlMedium: iconUrls?.medium ?? "",
    iconUrlHigh: iconUrls?.high ?? "",
    featuredRank: source.featured_rank,
    slug: source.slug ?? "",
    status: status ?? Status.Pending,
    createdAt: new Date(source.created_at),
    activatedAt: source.activated_at ? new Date(source.activated_at) : undefined,
    settledAt: source.settled_at ? new Date(source.settled_at) : undefined,
    resolutionBy: source.resolution_by ?? undefined,
    resolution: source.resolution,
    description: source.description,
    definition: definition as string,
    category: source.category,
    subcategory: source.subcategory,
    tags: source.tags ?? [],
    numOutcomes,
    depositAssets,
    depositAssetPairs,
    conditionalTokens,
    outcomes,
    orderbookPairs,
    orderbookIds: orderbookPairs.map((pair) => pair.orderbookId),
    tokenMetadata,
  };
}

/**
 * Derive unique base/quote deposit-asset pairs across the market's orderbook
 * pairs. Deduplicated by `(basePubkey, quotePubkey)`; orderbook pairs whose
 * base or quote deposit asset is not present in `depositAssets` are skipped.
 */
export function deriveDepositAssetPairs(
  depositAssets: DepositAsset[],
  orderbookPairs: OrderBookPair[],
): DepositAssetPair[] {
  const seen = new Map<string, DepositAssetPair>();

  for (const pair of orderbookPairs) {
    const base = depositAssets.find(
      (asset) => asset.depositAsset === pair.base.depositAsset,
    );
    const quote = depositAssets.find(
      (asset) => asset.depositAsset === pair.quote.depositAsset,
    );

    if (!base || !quote) continue;

    const key = `${base.depositAsset}|${quote.depositAsset}`;
    if (!seen.has(key)) {
      seen.set(key, {
        id: `${base.depositAsset}-${quote.depositAsset}`,
        base,
        quote,
      });
    }
  }

  return Array.from(seen.values());
}

export function tryMarketFromWire(source: MarketResponse): { market?: Market; error?: string } {
  try {
    return { market: marketFromWire(source) };
  } catch (error) {
    return { error: error instanceof Error ? error.message : String(error) };
  }
}
